packages/keyring-controller/src/KeyringController.ts

```diff
@@ -926,76 +928,83 @@ export class KeyringController extends BaseController<
+      serializedKeyrings.push(...this.#unsupportedKeyrings);
 
-    if (this.#cacheEncryptionKey) {
-      assertIsExportableKeyEncryptor(this.#encryptor);
+      let vault: string | undefined;
+      let newEncryptionKey: string | undefined;
 
-      if (encryptionKey) {
-        const key = await this.#encryptor.importKey(encryptionKey);
-        const vaultJSON = await this.#encryptor.encryptWithKey(
-          key,
-          serializedKeyrings,
-        );
-        vaultJSON.salt = encryptionSalt;
-        vault = JSON.stringify(vaultJSON);
-      } else if (this.#password) {
-        const { vault: newVault, exportedKeyString } =
-          await this.#encryptor.encryptWithDetail(
-            this.#password,
+      if (this.#cacheEncryptionKey) {
+        assertIsExportableKeyEncryptor(this.#encryptor);
+
+        if (encryptionKey) {
+          const key = await this.#encryptor.importKey(encryptionKey);
+          const vaultJSON = await this.#encryptor.encryptWithKey(
+            key,
             serializedKeyrings,
           );
-
-        vault = newVault;
-        newEncryptionKey = exportedKeyString;
-      }
-    } else {
-      if (typeof this.#password !== 'string') {
-        throw new TypeError(KeyringControllerError.WrongPasswordType);
+          vaultJSON.salt = encryptionSalt;
+          vault = JSON.stringify(vaultJSON);
+        } else if (this.#password) {
+          const { vault: newVault, exportedKeyString } =
+            await this.#encryptor.encryptWithDetail(
+              this.#password,
+              serializedKeyrings,
+            );
+
+          vault = newVault;
+          newEncryptionKey = exportedKeyString;
+        }
+      } else {
+        if (typeof this.#password !== 'string') {
+          throw new TypeError(KeyringControllerError.WrongPasswordType);
+        }
+        vault = await this.#encryptor.encrypt(
+          this.#password,
+          serializedKeyrings,
+        );
       }
-      vault = await this.#encryptor.encrypt(this.#password, serializedKeyrings);
-    }
-
-    if (!vault) {
-      throw new Error(KeyringControllerError.MissingVaultData);
-    }
 
-    this.update((state) => {
-      state.vault = vault;
-    });
+      if (!vault) {
+        throw new Error(KeyringControllerError.MissingVaultData);
+      }
 
-    // The keyring updates need to be announced before updating the encryptionKey
-    // so that the updated keyring gets propagated to the extension first.
-    // Not calling {@link updateKeyringsInState} results in the wrong account being selected
-    // in the extension.
-    await this.#updateKeyringsInState();
-    if (newEncryptionKey) {
+      // The keyring updates need to be announced before updating the encryptionKey
+      // so that the updated keyring gets propagated to the extension first.
+      // Not calling {@link updateKeyringsInState} results in the wrong account being selected
+      // in the extension.
```

Comment:
I'm not sure why updating `encryptionKey` and accounts together would result in the wrong account being selected in the extension. This comment might be outdated

Reply:
I suspect that the concern with this comment was updating the `encryptionKey` _before_ the `memState`. There was a case found in testing the MV3 integration where if they were updated separately but in the wrong order, it would set the account incorrectly (can't recall the exact details).

Much of that MV3 auto-login code has since been removed again because MV3 no longer shuts down the service worker mid-operation, so updating this state together definitely seems safe to me